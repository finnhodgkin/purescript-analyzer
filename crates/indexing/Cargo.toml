[package]
name = "indexing"
version = "0.1.0"
edition = "2021"

[dependencies]
fxhash = "0.2.1"
indexmap = "2.7.0"
la-arena = "0.3.1"
petgraph = "0.7.0"
rowan = "0.16.1"
rustc-hash = "2.1.0"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }

[dev-dependencies]
lexing = { path = "../lexing" }
parsing = { path = "../parsing" }
