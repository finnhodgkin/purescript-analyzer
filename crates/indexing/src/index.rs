use std::sync::Arc;

use la_arena::{Arena, ArenaMap, Idx};
use rustc_hash::FxHashMap;
use smol_str::SmolStr;

use crate::source::*;

#[derive(Debug, PartialEq, Eq)]
pub enum TermItem {
    ClassMember { id: ClassMemberId },
    Constructor { id: DataConstructorId },
    Derive { id: DeriveId },
    Foreign { id: ForeignValueId },
    Instance { id: InstanceId },
    Operator { id: InfixId },
    Value { signature: Option<ValueSignatureId>, equations: Vec<ValueEquationId> },
}

pub type TermItemId = Idx<TermItem>;

#[derive(Debug, PartialEq, Eq)]
pub enum TypeItem {
    Data {
        signature: Option<DataSignatureId>,
        equation: Option<DataEquationId>,
        role: Option<TypeRoleId>,
    },
    Newtype {
        signature: Option<NewtypeSignatureId>,
        equation: Option<NewtypeEquationId>,
        role: Option<TypeRoleId>,
    },
    Synonym {
        signature: Option<TypeSignatureId>,
        equation: Option<TypeEquationId>,
    },
    Class {
        signature: Option<ClassSignatureId>,
        declaration: Option<ClassDeclarationId>,
    },
    Foreign {
        id: ForeignDataId,
    },
    Operator {
        id: InfixId,
    },
}

pub type TypeItemId = Idx<TypeItem>;

#[derive(Debug, PartialEq, Eq)]
pub enum ImplicitItems {
    Everything,
    Enumerated(Arc<[SmolStr]>),
}

pub type ImportedTerms = FxHashMap<SmolStr, ImportItemId>;
pub type ImportedTypes = FxHashMap<SmolStr, (ImportItemId, Option<ImplicitItems>)>;

#[derive(Debug, PartialEq, Eq)]
pub struct ImportedItems {
    pub terms: ImportedTerms,
    pub types: ImportedTypes,
    pub exported: bool,
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct Index {
    term_item: Arena<TermItem>,
    type_item: Arena<TypeItem>,
    term_export: ArenaMap<TermItemId, ExportItemId>,
    type_export: ArenaMap<TypeItemId, ExportItemId>,
    alias_nominal: FxHashMap<SmolStr, Vec<ImportId>>,
    imported_items: FxHashMap<ImportId, ImportedItems>,
    term_nominal: FxHashMap<SmolStr, TermItemId>,
    type_nominal: FxHashMap<SmolStr, TypeItemId>,
}

impl std::ops::Index<TermItemId> for Index {
    type Output = TermItem;

    fn index(&self, id: TermItemId) -> &TermItem {
        &self.term_item[id]
    }
}

impl std::ops::Index<TypeItemId> for Index {
    type Output = TypeItem;

    fn index(&self, id: TypeItemId) -> &TypeItem {
        &self.type_item[id]
    }
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct Relational {
    data: Vec<(TypeItemId, TermItemId)>,
    class: Vec<(TypeItemId, TermItemId)>,
    chain: Vec<(InstanceChainId, InstanceId)>,
    instance: Vec<(InstanceId, InstanceMemberId)>,
}

impl Index {
    pub(crate) fn insert_term_item(&mut self, k: Option<SmolStr>, v: TermItem) -> TermItemId {
        let id = self.term_item.alloc(v);
        let Some(k) = k else { return id };
        self.term_nominal.insert(k, id);
        id
    }

    pub(crate) fn term_item_mut(&mut self, k: &str) -> Option<(&mut TermItem, TermItemId)> {
        let &id = self.term_nominal.get(k)?;
        Some((&mut self.term_item[id], id))
    }

    pub(crate) fn export_term_item(&mut self, item_id: TermItemId, id: ExportItemId) {
        self.term_export.insert(item_id, id);
    }

    pub(crate) fn term_item_export(&self, id: TermItemId) -> Option<ExportItemId> {
        self.term_export.get(id).copied()
    }

    pub(crate) fn insert_type_item(&mut self, k: Option<SmolStr>, v: TypeItem) -> TypeItemId {
        let id = self.type_item.alloc(v);
        let Some(k) = k else { return id };
        self.type_nominal.insert(k, id);
        id
    }

    pub(crate) fn type_item_mut(&mut self, k: &str) -> Option<(&mut TypeItem, TypeItemId)> {
        let &id = self.type_nominal.get(k)?;
        Some((&mut self.type_item[id], id))
    }

    pub(crate) fn export_type_item(&mut self, item_id: TypeItemId, id: ExportItemId) {
        self.type_export.insert(item_id, id);
    }

    pub(crate) fn type_item_export(&self, id: TypeItemId) -> Option<ExportItemId> {
        self.type_export.get(id).copied()
    }

    pub(crate) fn insert_import_alias(&mut self, k: SmolStr, v: ImportId) {
        self.alias_nominal.entry(k).or_default().push(v);
    }

    pub(crate) fn export_import_alias(&mut self, k: &str) {
        if let Some(imports) = self.alias_nominal.get(k) {
            for import in imports {
                if let Some(items) = self.imported_items.get_mut(import) {
                    items.exported = true;
                }
            }
        }
    }

    pub(crate) fn insert_imported_items(&mut self, k: ImportId, v: ImportedItems) {
        self.imported_items.insert(k, v);
    }

    pub fn lookup_import_alias(&self, k: &str) -> Option<&[ImportId]> {
        Some(self.alias_nominal.get(k)?)
    }

    pub fn iter_term_item(&self) -> impl Iterator<Item = (TermItemId, &TermItem)> {
        self.term_item.iter()
    }

    pub fn iter_type_item(&self) -> impl Iterator<Item = (TypeItemId, &TypeItem)> {
        self.type_item.iter()
    }
}

impl Relational {
    pub(crate) fn insert_data_relation(&mut self, type_id: TypeItemId, term_id: TermItemId) {
        self.data.push((type_id, term_id));
    }

    pub(crate) fn insert_class_relation(&mut self, type_id: TypeItemId, term_id: TermItemId) {
        self.class.push((type_id, term_id));
    }

    pub(crate) fn insert_chain_relation(&mut self, c_id: InstanceChainId, i_id: InstanceId) {
        self.chain.push((c_id, i_id));
    }

    pub(crate) fn insert_instance_relation(&mut self, i_id: InstanceId, m_id: InstanceMemberId) {
        self.instance.push((i_id, m_id));
    }

    pub fn constructors_of(&self, id: TypeItemId) -> impl Iterator<Item = TermItemId> + '_ {
        self.data.iter().filter_map(
            move |(type_id, term_id)| {
                if id == *type_id {
                    Some(*term_id)
                } else {
                    None
                }
            },
        )
    }

    pub fn class_members_of(&self, id: TypeItemId) -> impl Iterator<Item = TermItemId> + '_ {
        self.class.iter().filter_map(
            move |(type_id, term_id)| {
                if id == *type_id {
                    Some(*term_id)
                } else {
                    None
                }
            },
        )
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum ItemKind {
    DataSignature(DataSignatureId),
    DataEquation(DataEquationId),

    NewtypeSignature(NewtypeSignatureId),
    NewtypeEquation(NewtypeEquationId),

    Constructor(DataConstructorId),

    SynonymSignature(TypeSignatureId),
    SynonymEquation(TypeEquationId),

    ClassSignature(ClassSignatureId),
    ClassDeclaration(ClassDeclarationId),
    ClassMember(ClassMemberId),

    ValueSignature(ValueSignatureId),
    ValueEquation(ValueEquationId),

    ForeignData(ForeignDataId),
    ForeignValue(ForeignValueId),
    Operator(InfixId),
}

#[derive(Debug, PartialEq, Eq)]
pub enum ExistingKind {
    Term(TermItemId),
    Type(TypeItemId),
}

#[derive(Debug, PartialEq, Eq)]
pub enum IndexError {
    DuplicateItem { kind: ItemKind, existing: ExistingKind },
    MismatchedItem { kind: ItemKind, existing: ExistingKind },
    InvalidRole { id: TypeRoleId, existing: Option<TypeItemId> },
    InvalidExport { id: ExportItemId },
    DuplicateExport { id: ExportItemId, existing: ExportItemId },
    DuplicateImport { id: ImportItemId, existing: ImportItemId },
}
