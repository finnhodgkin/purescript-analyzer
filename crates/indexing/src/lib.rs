mod algorithm;

use fxhash::{FxBuildHasher, FxHashMap};
use la_arena::{Arena, ArenaMap, Idx};
use petgraph::{prelude::GraphMap, Directed};
use smol_str::SmolStr;
use syntax::{cst, SyntaxNode, SyntaxNodePtr};

pub type DeclarationId = Idx<cst::Declaration>;

#[derive(Debug, PartialEq, Eq)]
pub struct ClassGroup {
    pub signature: Option<DeclarationId>,
    pub declaration: Option<DeclarationId>,
}

#[derive(Debug, Default)]
pub struct ClassIndex {
    pub by_type: FxHashMap<SmolStr, ClassGroup>,
    pub by_member: FxHashMap<SmolStr, DeclarationId>,
    pub statement_graph: GraphMap<DeclarationId, (), Directed, FxBuildHasher>,
}

#[derive(Debug, Default)]
pub struct InstanceIndex {
    pub by_type: FxHashMap<SmolStr, Vec<DeclarationId>>,
    pub by_name: FxHashMap<SmolStr, DeclarationId>,
    pub instance_graph: GraphMap<DeclarationId, (), Directed, FxBuildHasher>,
    pub statement_graph: GraphMap<DeclarationId, (), Directed, FxBuildHasher>,
}

#[derive(Debug, PartialEq, Eq)]
pub struct ValueGroup {
    pub signature: Option<DeclarationId>,
    pub equations: Vec<DeclarationId>,
}

pub type ValueIndex = FxHashMap<SmolStr, ValueGroup>;

#[derive(Debug, Default)]
pub struct FullIndexingResult {
    arena: Arena<cst::Declaration>,
    pub pointer: ArenaMap<DeclarationId, SyntaxNodePtr>,
    pub class: ClassIndex,
    pub instance: InstanceIndex,
    pub value: ValueIndex,
    pub errors: Vec<IndexingError>,
}

#[derive(Debug, PartialEq, Eq)]
pub enum IndexingError {
    SignatureConflict { existing: DeclarationId, duplicate: DeclarationId },
    SignatureIsLate { equation: DeclarationId, signature: DeclarationId },
    DeclarationConflict { existing: DeclarationId, duplicate: DeclarationId },
}

pub fn index(node: SyntaxNode) -> FullIndexingResult {
    let mut module_map = FullIndexingResult::default();
    algorithm::index_module(&mut module_map, node);
    module_map
}
