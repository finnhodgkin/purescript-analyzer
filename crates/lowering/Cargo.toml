[package]
name = "lowering"
version = "0.1.0"
edition = "2021"

[dependencies]
id = { version = "0.1.0", path = "../id" }
indexing = { version = "0.1.0", path = "../indexing" }
indexmap = "2.7.1"
lexing = { version = "0.1.0", path = "../lexing" }
parsing = { version = "0.1.0", path = "../parsing" }
rowan = "0.16.1"
rustc-hash = "2.1.0"
syntax = { version = "0.1.0", path = "../syntax" }
