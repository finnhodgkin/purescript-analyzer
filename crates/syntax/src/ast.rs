#[macro_use]
mod macros;

mod binder;
mod binding;
mod common;
mod declaration;
mod module;
mod name;

pub use binder::*;
pub use binding::*;
pub use common::*;
pub use declaration::*;
pub use module::*;
pub use name::*;
