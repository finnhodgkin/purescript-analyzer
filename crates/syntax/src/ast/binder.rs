use rowan::ast::AstNode;

use crate::SyntaxToken;

use super::{Name, OneOrMore, QualifiedName};

_create_ast_v!(
    Binder,
    ConstructorBinder(ConstructorBinder),
    LiteralBinder(LiteralBinder),
    NegativeBinder(NegativeBinder),
    ParenthesizedBinder(ParenthesizedBinder),
    TypedBinder(TypedBinder),
    VariableBinder(VariableBinder),
    WildcardBinder(WildcardBinder)
);

impl ConstructorBinder {
    pub fn qualified_name(&self) -> Option<QualifiedName> {
        QualifiedName::cast(self.node.first_child()?)
    }

    pub fn fields(&self) -> Option<OneOrMore<Binder>> {
        OneOrMore::cast(self.node.children().nth(1)?)
    }
}

impl ParenthesizedBinder {
    pub fn binder(&self) -> Option<Binder> {
        Binder::cast(self.node.first_child()?)
    }
}

impl NegativeBinder {
    pub fn minus(&self) -> Option<SyntaxToken> {
        self.node.first_token()
    }

    pub fn literal(&self) -> Option<LiteralBinder> {
        LiteralBinder::cast(self.node.first_child()?)
    }
}

impl VariableBinder {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }
}
