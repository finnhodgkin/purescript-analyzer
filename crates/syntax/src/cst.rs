use rowan::ast::{AstChildren, AstNode};

use crate::{SyntaxKind, SyntaxToken};

#[macro_use]
mod macros;

create_cst_struct!(Module, ModuleHeader, ExportList);

create_cst_enum!(
    ExportItem
        | ExportValue
        | ExportClass
        | ExportType
        | ExportOperator
        | ExportTypeOperator
        | ExportModule
);

create_cst_struct!(ModuleImports, ImportStatement, ImportList, ImportAlias);

create_cst_enum!(
    ImportItem // Comment to fix formatting :P
        | ImportValue
        | ImportClass
        | ImportType
        | ImportOperator
        | ImportTypeOperator
);

create_cst_enum!(TypeItems | TypeItemsAll | TypeItemsList);

create_cst_struct!(ModuleStatements);

create_cst_enum!(
    Declaration
        | ValueSignature
        | ValueEquation
        | InfixDeclaration
        | TypeRoleDeclaration
        | InstanceChain
        | InstanceDeclaration
        | InstanceSignatureStatement
        | InstanceEquationStatement
        | TypeSynonymSignature
        | TypeSynonymEquation
        | ClassSignature
        | ClassDeclaration
        | ClassMemberStatement
        | ForeignImportDataDeclaration
        | ForeignImportValueDeclaration
        | NewtypeSignature
        | NewtypeEquation
        | DataSignature
        | DataEquation
        | DeriveDeclaration
);

associated_declarations!(
    ClassDeclaration where ClassMemberStatement
);

associated_declarations!(
    InstanceDeclaration where InstanceSignatureStatement | InstanceEquationStatement
);

create_cst_struct!(
    TypeRole,
    ClassConstraints,
    ClassHead,
    ClassFunctionalDependencies,
    ClassStatements
);

create_cst_enum!(
    FunctionalDependency | FunctionalDependencyDetermined | FunctionalDependencyDetermines
);

create_cst_struct!(InstanceName, InstanceConstraints, InstanceHead, InstanceStatements);

create_cst_struct!(DataConstructor);

create_cst_enum!(
    Type // Comment to fix formatting :P
        | TypeApplicationChain
        | TypeArrow
        | TypeConstrained
        | TypeConstructor
        | TypeForall
        | TypeHole
        | TypeInteger
        | TypeKinded
        | TypeOperator
        | TypeOperatorChain
        | TypeString
        | TypeVariable
        | TypeVariableBinding
        | TypeWildcard
        | TypeRecord
        | TypeRow
        | TypeParenthesized
);

create_cst_enum!(
    Binder
        | BinderTyped
        | BinderOperatorChain
        | BinderInteger
        | BinderNumber
        | BinderConstructor
        | BinderVariable
        | BinderNamed
        | BinderWildcard
        | BinderString
        | BinderChar
        | BinderTrue
        | BinderFalse
        | BinderArray
        | BinderRecord
        | BinderParenthesized
);

create_cst_struct!(
    FunctionBinders,
    Unconditional,
    WhereExpression,
    Conditionals,
    PatternGuarded,
    PatternGuards
);

create_cst_enum!(PatternGuard | PatternGuardBinder | PatternGuardExpression);

create_cst_enum!(
    Expression
        | ExpressionTyped
        | ExpressionOperatorChain
        | ExpressionInfixChain
        | ExpressionTick
        | ExpressionNegate
        | ExpressionApplicationChain
        | ExpressionTypeArgument
        | ExpressionTermArgument
        | ExpressionIfThenElse
        | ExpressionLetIn
        | ExpressionLambda
        | ExpressionCaseOf
        | ExpressionDo
        | ExpressionAdo
        | ExpressionConstructor
        | ExpressionVariable
        | ExpressionOperatorName
        | ExpressionSection
        | ExpressionHole
        | ExpressionString
        | ExpressionChar
        | ExpressionTrue
        | ExpressionFalse
        | ExpressionInteger
        | ExpressionNumber
        | ExpressionArray
        | ExpressionRecord
        | ExpressionParenthesized
        | ExpressionRecordAccess
        | ExpressionRecordUpdate
);

create_cst_struct!(CaseTrunk, CaseBranches, CaseBranchBinders, CaseBranch);

create_cst_enum!(DoStatement | DoStatementBind | DoStatementLet | DoStatementDiscard);

create_cst_enum!(LetBinding | LetBindingPattern | LetBindingSignature | LetBindingEquation);

create_cst_enum!(RecordItem | RecordField | RecordPun);

create_cst_struct!(RecordUpdates);

create_cst_enum!(RecordUpdate | RecordUpdateLeaf | RecordUpdateBranch);

impl Module {
    pub fn statements(&self) -> Option<ModuleStatements> {
        self.syntax().children().find_map(ModuleStatements::cast)
    }
}

impl ModuleStatements {
    pub fn children(&self) -> impl Iterator<Item = Declaration> {
        self.syntax().children().filter_map(Declaration::cast)
    }
}

impl InstanceChain {
    pub fn instance_declarations(&self) -> AstChildren<InstanceDeclaration> {
        rowan::ast::support::children(self.syntax())
    }
}

impl InstanceDeclaration {
    pub fn instance_head(&self) -> Option<InstanceHead> {
        rowan::ast::support::child(self.syntax())
    }

    pub fn instance_name(&self) -> Option<InstanceName> {
        rowan::ast::support::child(self.syntax())
    }

    pub fn instance_statements(&self) -> Option<InstanceStatements> {
        rowan::ast::support::child(self.syntax())
    }
}

impl InstanceName {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}

impl InstanceHead {
    pub fn type_name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::UPPER)
    }
}

impl InstanceStatements {
    pub fn children(&self) -> AstChildren<InstanceDeclarationStatement> {
        rowan::ast::support::children(self.syntax())
    }
}

impl ValueSignature {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}

impl ValueEquation {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}
