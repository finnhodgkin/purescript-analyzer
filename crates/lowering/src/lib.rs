use indexing::IndexingResult;
use syntax::cst;

mod algorithm;
mod concrete;
mod sourcemap;

pub use concrete::*;
pub use sourcemap::*;

pub fn lower(module: &cst::Module, index: &IndexingResult) {
    algorithm::lower_module(module, &index);
}
