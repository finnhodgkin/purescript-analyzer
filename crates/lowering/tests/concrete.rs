use indexing::IndexingResult;
use lowering::LoweringResult;
use rowan::ast::AstNode;
use syntax::cst;

fn lower_declaration<'s>(source: impl AsRef<[&'s str]>) -> (IndexingResult, LoweringResult) {
    let source = format!("module Main where\n\n{}", source.as_ref().join("\n"));

    let lexed = lexing::lex(&source);
    let tokens = lexing::layout(&lexed);

    let (node, _) = parsing::parse(&lexed, &tokens);
    let module = cst::Module::cast(node).unwrap();

    let (index, _) = indexing::index(&module);
    let lower = lowering::lower(&module, &index);

    (index, lower)
}

#[test]
fn lower_value_equation() {
    let (index, lower) = lower_declaration(["id :: forall a. a -> a", "id a = a", "id b = b"]);

    let (id, _, _) = index.nominal.lookup_expr_item("id").unwrap();
    let id = lower.lowering_map.expr_item.get(&id).unwrap();

    insta::assert_debug_snapshot!(id);
}

#[test]
fn lower_value_equation_partial() {
    let (index, lower) = lower_declaration(["id :: forall a. a -> a", "id a = ;", "id b = b"]);

    let (id, _, _) = index.nominal.lookup_expr_item("id").unwrap();
    let id = lower.lowering_map.expr_item.get(&id).unwrap();

    insta::assert_debug_snapshot!(id);
}

#[test]
fn lower_expression_typed() {
    let (_, lower) = lower_declaration(["four = 4 :: Int", "five = 5 ::"]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_operator_chain() {
    let (_, lower) = lower_declaration(["plus = 1 + 2 + 3 + 4 + 5"]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_infix_chain() {
    let (_, lower) = lower_declaration(["chain = a `for` b `for` c"]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_negate() {
    let (_, lower) = lower_declaration(["negate = -1"]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_application_chain() {
    let (_, lower) = lower_declaration(["negate = f @a @b a b"]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_if_then_else() {
    let (_, lower) = lower_declaration(["ifThenElse = if a then b else c"]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_let_in() {
    let (_, lower) = lower_declaration([
        "letIn =",
        "  let",
        "    life :: Int",
        "    life = 42",
        "",
        "    42 = value",
        "      where",
        "      value = 42",
        "",
        "  in",
        "    life",
    ]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_lambda() {
    let (_, lower) = lower_declaration(["id = \\a -> a"]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_conditionals() {
    let (_, lower) = lower_declaration([
        "id a",
        "  | b <- c",
        "  , d <- e",
        "  , f <- g = 0",
        "  | true = 1",
        "  | false = 1",
    ]);
    insta::assert_debug_snapshot!(&lower.lowering_map.expr_item);
}

#[test]
fn lower_expression_case_of() {
    let (_, lower) = lower_declaration([
        "id = case _ of",
        "  a | b <- c",
        "    , d <- e",
        "    , f <- g -> 0",
        "    | true -> 1",
        "    | false -> 1",
    ]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_do() {
    let (_, lower) = lower_declaration([
        "main = Hooks.do",
        "  pure unit",
        "  _ <- pure unit",
        "  let",
        "    a = pure b",
        "  pure unit",
    ]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_ado() {
    let (_, lower) = lower_declaration([
        "main = Hooks.ado",
        "  _ <- action",
        "  let",
        "    id a = a",
        "  _ <- action",
        "  in result",
    ]);
    insta::assert_debug_snapshot!(&lower.source_map);
}

#[test]
fn lower_expression_name() {
    let (_, lower) = lower_declaration(["id = F.Id F.id F.(+)"]);
    insta::assert_debug_snapshot!(&lower.source_map);
}
