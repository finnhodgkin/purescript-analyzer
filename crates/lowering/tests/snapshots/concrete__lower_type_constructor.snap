---
source: crates/lowering/tests/concrete.rs
expression: "&lower.source_map"
snapshot_kind: text
---
SourceMap {
    types: {
        AstPtr {
            raw: SyntaxNodePtr {
                kind: TypeConstructor,
                range: 24..30,
            },
        }: Constructor {
            qualifier: Some(
                "Id.",
            ),
            name: Some(
                "Id",
            ),
        },
    },
    binders: {},
    expressions: {},
    let_bindings: {},
    let_bindings_grouped: [],
    do_statements: {},
}
